In [3]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
from sklearn import ensemble
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import roc_auc_score

import numpy as np
from sklearn import datasets

In [4]:
# load data
cal = datasets.fetch_california_housing()
data = cal['data']
targets = cal['target']

array([ 8.30140000e+00,  2.10000000e+01,  6.23813708e+00,  9.71880492e-01,
        2.40100000e+03,  2.10984183e+00,  3.78600000e+01, -1.22220000e+02])

# Task 1

In [8]:
# 1a
clf = LinearRegression()
scores = cross_val_score(clf, data, targets, cv=5)
print('Linear Regression R^2 Scores', scores)
print('Linear Regression Mean R^2 Score', np.mean(scores))
print()

clf = ensemble.GradientBoostingRegressor()
scores = cross_val_score(clf, data, targets, cv=5)
print('Boosting R^2 Scores', scores)
print('Boosting Scores Mean R^2 Score', np.mean(scores))

Linear Regression R^2 Scores [0.54866323 0.46820691 0.55078434 0.53698703 0.66051406]
Linear Regression Mean R^2 Score 0.5530311140279233

Boosting R^2 Scores [0.60256286 0.69877396 0.7180343  0.65023363 0.67979733]
Boosting Scores Mean R^2 Score 0.6698804157532645


In [4]:
# 1b
tuned_parameters = [{'max_depth': [3, 10],
                     'n_estimators': [50, 100],
                     'learning_rate': [0.01, 0.1]}]
clf = ensemble.GradientBoostingRegressor()
clf = GridSearchCV(clf, tuned_parameters)
clf.fit(data, targets)

print("Scores for parameter grid search:")
print()
means = clf.cv_results_['mean_test_score']
stds = clf.cv_results_['std_test_score']
for mean, std, params in zip(means, stds, clf.cv_results_['params']):
    print("%0.3f (+/-%0.03f) for %r"
          % (mean, std * 2, params))

Scores for parameter grid search:

0.280 (+/-0.105) for {'learning_rate': 0.01, 'max_depth': 3, 'n_estimators': 50}
0.435 (+/-0.121) for {'learning_rate': 0.01, 'max_depth': 3, 'n_estimators': 100}
0.374 (+/-0.053) for {'learning_rate': 0.01, 'max_depth': 10, 'n_estimators': 50}
0.542 (+/-0.077) for {'learning_rate': 0.01, 'max_depth': 10, 'n_estimators': 100}
0.633 (+/-0.094) for {'learning_rate': 0.1, 'max_depth': 3, 'n_estimators': 50}
0.670 (+/-0.081) for {'learning_rate': 0.1, 'max_depth': 3, 'n_estimators': 100}
0.652 (+/-0.139) for {'learning_rate': 0.1, 'max_depth': 10, 'n_estimators': 50}
0.659 (+/-0.129) for {'learning_rate': 0.1, 'max_depth': 10, 'n_estimators': 100}


**1c) Briefly discuss the performance and summarize your findings.**

When running linear regression and gradient boosting using 5-fold cross validation, we got the following R^2 Score results:

    Linear Regression Mean Score = 0.5530311140279233
    Boosting Scores Mean Score = 0.6698681752149087

Overall the scores are not great. R^2 measures the goodness-of-fit of the model on the data. R^2 closer to 1 is better. The boosting score is significantly greater than the linear regressor which makes sense.


For 1b) we tested all of the following possible permutations:

    'max_depth': [3, 10],
    'n_estimators': [50, 100],
    'learning_rate': [0.01, 0.1]
    
Our parameter grid search yielded interesting results. None of the parameter combinations were convincingly better than the default parameters. In fact, increasing the max_depth and decreasing the learning rate were very detrimental to the R^2 score.


# Task 2

In [10]:
new_targets = np.array([x>2 for x in targets])

In [11]:
# 2a
clf = LogisticRegression()
scores = cross_val_score(clf, data, new_targets, cv=5)
print('Logistic Regression Scores', scores)
print('Logistic Regression Mean Score', np.mean(scores))
print()

clf = ensemble.GradientBoostingClassifier()
scores = cross_val_score(clf, data, new_targets, cv=5)
print('Boosting Classifier Scores', scores)
print('Boosting Classifier Mean Score', np.mean(scores))

/Users/David/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/Users/David/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/Users/David/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/Users/David/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/Users/David/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will 

Logistic Regression Scores [0.80988133 0.79796512 0.77616279 0.74612403 0.82481221]
Logistic Regression Mean Score 0.7909890954886174

Boosting Classifier Scores [0.79099055 0.75436047 0.80741279 0.75314922 0.82650836]
Boosting Classifier Mean Score 0.786484278963419


In [7]:
# 2b
tuned_parameters = [{'max_depth': [3, 5],
                     'n_estimators': [100, 200],
                     'learning_rate': [0.1, 0.5]}]
clf = ensemble.GradientBoostingClassifier()
clf = GridSearchCV(clf, tuned_parameters)
clf.fit(data, new_targets)

print("Scores for parameter grid search:")
print()
means = clf.cv_results_['mean_test_score']
stds = clf.cv_results_['std_test_score']
for mean, std, params in zip(means, stds, clf.cv_results_['params']):
    print("%0.3f (+/-%0.03f) for %r"
          % (mean, std * 2, params))

Scores for parameter grid search:

0.788 (+/-0.059) for {'learning_rate': 0.1, 'max_depth': 3, 'n_estimators': 100}
0.774 (+/-0.095) for {'learning_rate': 0.1, 'max_depth': 3, 'n_estimators': 200}
0.765 (+/-0.100) for {'learning_rate': 0.1, 'max_depth': 5, 'n_estimators': 100}
0.752 (+/-0.125) for {'learning_rate': 0.1, 'max_depth': 5, 'n_estimators': 200}
0.751 (+/-0.130) for {'learning_rate': 0.5, 'max_depth': 3, 'n_estimators': 100}
0.748 (+/-0.139) for {'learning_rate': 0.5, 'max_depth': 3, 'n_estimators': 200}
0.741 (+/-0.125) for {'learning_rate': 0.5, 'max_depth': 5, 'n_estimators': 100}
0.742 (+/-0.136) for {'learning_rate': 0.5, 'max_depth': 5, 'n_estimators': 200}


In [8]:
# 2c
clf = LogisticRegression()
scores = cross_val_score(clf, data, new_targets, cv=5, scoring='roc_auc')
print('Logistic Regression ROC AUC scores', scores)


clf = ensemble.GradientBoostingClassifier()
scores = cross_val_score(clf, data, new_targets, cv=5, scoring='roc_auc')
print('Boosting Classifier ROC AUC scores', scores)

/home/david/.local/lib/python3.6/site-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)
/home/david/.local/lib/python3.6/site-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logi

Logistic Regression ROC AUC scores [0.87144465 0.88084073 0.87614029 0.81245699 0.89996528]
Boosting Classifier ROC AUC scores [0.8848087  0.84739915 0.90709042 0.9047687  0.91558129]


2d)
